homeassistant/components/tag/__init__.py

```diff
@@ -90,21 +125,231 @@ async def _update_data(self, item: dict, update_data: dict) -> dict:
         # make last_scanned JSON serializeable
         if LAST_SCANNED in update_data:
             data[LAST_SCANNED] = data[LAST_SCANNED].isoformat()
+        if name := data.get(CONF_NAME):
+            entity_id = self.entity_registry.async_get_entity_id(
+                DOMAIN, DOMAIN, data[TAG_ID]
+            )
+            if entity_id := self.entity_registry.async_get_entity_id(
+                DOMAIN, DOMAIN, data[TAG_ID]
+            ):
+                self.entity_registry.async_update_entity(entity_id, name=name)
+            else:
+                self.entity_registry.async_get_or_create(
+                    DOMAIN,
+                    DOMAIN,
+                    data[TAG_ID],
+                    original_name=data.get(CONF_NAME, DEFAULT_NAME),
+                    suggested_object_id=slugify(data.get(CONF_NAME, DEFAULT_NAME)),
+                )
+            data.pop(CONF_NAME)
+
         return data
 
 
+class TagDictStorageCollectionWebsocket(
+    collection.StorageCollectionWebsocket[collection.DictStorageCollection]
+):
+    """Class to expose tag storage collection management over websocket."""
+
+    def __init__(
+        self,
+        storage_collection: collection.DictStorageCollection,
+        api_prefix: str,
+        model_name: str,
+        create_schema: ConfigType,
+        update_schema: ConfigType,
+    ) -> None:
+        """Initialize a websocket for tag."""
+        super().__init__(
+            storage_collection, api_prefix, model_name, create_schema, update_schema
+        )
+        self.entity_registry = er.async_get(storage_collection.hass)
+
+    @callback
+    def ws_list_item(
+        self, hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
+    ) -> None:
+        """List items specifically for tag.
+
+        Provides name from entity_registry instead of storage collection.
+        """
+        tag_items = self.storage_collection.async_items()
+        for item in tag_items:
+            if (
+                entity_id := self.entity_registry.async_get_entity_id(
+                    DOMAIN, DOMAIN, item[TAG_ID]
+                )
+            ) and (entity := self.entity_registry.async_get(entity_id)):
+                item[CONF_NAME] = entity.name or entity.original_name
+        if _LOGGER.isEnabledFor(logging.DEBUG):
+            _LOGGER.debug("Listing tags %s", tag_items)
+        connection.send_result(msg["id"], tag_items)
+
+    async def ws_create_item(
+        self, hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
+    ) -> None:
```

Comment:
Why do we need to override this method? The entity registry is accessed by `TagStorageCollection._process_create_data` why do we mess with it here too?

Reply:
We need to modify the message to include the name from entity registry or it comes back to frontend as blank until the webpage is refreshed (when it get's it from ws_list instead.
I think this would be preferred instead of modifying the frontend.